src/lib/transaction_logic/test/transaction_logic/transaction_logic.ml

```diff
@@ -0,0 +1,149 @@
+      ; common =
+          (let open Signed_command.Payload.Common.Poly in
+          { fee = fee'
+          ; fee_payer_pk = fee_payer'.pk
+          ; nonce = fee_payer'.nonce
+          ; valid_until = valid
+          ; memo = Signed_command_memo.dummy
+          })
+      }
+  in
+  Command
+    (User_command.Signed_command
+       Signed_command.Poly.{ payload; signer; signature = Signature.dummy } )
+
+let simple_payment () =
+  Quickcheck.test
+    (let open Quickcheck.Generator.Let_syntax in
+    let%bind sender =
+      Test_account.gen_constrained_balance ~min:Balance.one ()
+    in
+    let%bind receiver = Test_account.gen_empty in
+    let%bind amount =
+      Amount.(gen_incl one @@ Balance.to_amount sender.balance)
+    in
+    let max_fee =
+      let open Amount in
+      Balance.to_amount sender.balance - amount
+      |> Option.value_map ~f:to_fee ~default:Fee.zero
+    in
+    let%bind fee = Fee.(gen_incl zero max_fee) in
+    let%map global_slot = Global_slot.gen in
+    let txn = signed_command ~fee ~sender ~receiver amount in
+    (global_slot, [ sender; receiver ], [ txn ]))
+    ~f:(fun (global_slot, accounts, txns) ->
+      let txn_state_view = protocol_state ~global_slot accounts in
+      let ledger =
+        match test_ledger accounts with Ok l -> l | Error _ -> assert false
+      in
+      [%test_pred: Transaction_applied.t list Or_error.t] Or_error.is_ok
+        (Transaction_logic.apply_transactions ~constraint_constants ~global_slot
+           ~txn_state_view ledger txns ) )
+
+let balance_to_fee b = Balance.to_amount b |> Amount.to_fee
+
+let simple_payment_signer_different_from_fee_payer () =
+  Quickcheck.test
+    (let open Quickcheck.Generator.Let_syntax in
+    let%bind sender =
+      Test_account.gen_constrained_balance ~min:Balance.one ()
+    in
+    let%bind amount =
+      Amount.(gen_incl one @@ Balance.to_amount sender.balance)
+    in
+    let max_rcv_balance =
+      Balance.(sub_amount max_int amount) |> Option.value ~default:Balance.zero
+    in
+    let%bind receiver =
+      Test_account.gen_constrained_balance ~max:max_rcv_balance ()
+    in
+    let%bind fee = Fee.(gen_incl zero @@ balance_to_fee receiver.balance) in
+    let%map global_slot = Global_slot.gen in
+    let txn =
+      signed_command ~signer:sender ~fee_payer:receiver ~fee ~sender ~receiver
+        amount
+    in
+    (global_slot, [ sender; receiver ], [ txn ]))
+    ~f:(fun (global_slot, accounts, txns) ->
+      let txn_state_view = protocol_state ~global_slot accounts in
+      let ledger =
+        match test_ledger accounts with Ok l -> l | Error _ -> assert false
+      in
+      expect_failure
+        ~error:
+          "Cannot pay fees from a public key that did not sign the transaction"
+        (Transaction_logic.apply_transactions ~constraint_constants ~global_slot
+           ~txn_state_view ledger txns ) )
+
+let simple_payment_signer_different_from_sender () =
+  Quickcheck.test
```

Comment:
Could we extract a helper for all three simple payment tests to reduce code duplication?  They seem to differ only in last few lines.

Reply:
Done in 6543ba6c23. Code isn't much shorter, because of the nuances cause by different fee payer's in all three tests, but still I think it looks better. Thanks!